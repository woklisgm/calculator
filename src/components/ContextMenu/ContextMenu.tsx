import React from 'react';

import styles from './context-menu.module.css';
import { useActions } from '../../hooks/useActions';

function ContextMenu({x, y}: {x: number, y: number}) {
	const {pastExpression, clearExpression} = useActions();

	const handlePaste = () => {
		navigator.clipboard.readText().then(text => {
				text && pastExpression(text);
			})
			.catch(err => {
				console.log('Error read from buffer');
			});
	}

	const handleClear = () => {
		clearExpression();
	}

	return (
		<div className={styles.contextmenu} style={{left: x, top: y}}>
			<div 
				data-testid="past"
				onClick={handlePaste}
				className={styles.item} 
			>
				Вставить
			</div>
			<div
				data-testid="clear" 
				onClick={handleClear}
				className={styles.item} 
			>
				Очистить
			</div>
			<div className={styles.item}>
				...
			</div>
		</div>
	)
}

export {ContextMenu};