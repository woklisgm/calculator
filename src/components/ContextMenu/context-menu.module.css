.contextmenu {
	padding-top: 10px;
	padding-bottom: 10px;
	font-family: 'Geometria';
	position: absolute;
	width: 150px;
	border-radius: 10px;
	background-color: rgba(255, 255, 255, 0.9);
	box-shadow: 0px 10.2677px 19.7841px rgba(0, 0, 0, 0.175), 0px 3.71362px 7.1555px rgba(0, 0, 0, 0.121944);
}

.item {
	height: 30px;
	cursor: pointer;
	font-size: 18px;
	padding-left: 10px;
}

.item:hover {
	height: 30px;
	padding-left: 10px;
	background-color: rgb(230, 240, 255);
}

.item:last-child {
	padding-left: 0;
	text-align: center;
}

.item:last-child:hover {
	cursor: default;
	background-color: rgba(255, 255, 255, 0.0);
}