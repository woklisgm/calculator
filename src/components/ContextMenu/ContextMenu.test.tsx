/* eslint-disable testing-library/no-debugging-utils */
import React from 'react';
import renderer  from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';

import { ContextMenu } from './ContextMenu';
import { renderWithRedux } from '../../tests/helpers/renderWithRedux';

const mockPastExpression = jest.fn();
const mockClearExpression = jest.fn();

jest.mock('../../hooks/useActions', () => ({
	useActions: () => ({
		pastExpression: mockPastExpression,
		clearExpression: mockClearExpression
	})
}));

Object.assign(navigator, {
  clipboard: {
    readText: () => {},
  },
});

beforeEach(() => {
	jest.clearAllMocks();
})

describe('Keyboard tests', () => {
	test('snapshot', () => {
		const treeApp = renderer.create(renderWithRedux(<ContextMenu x={0} y={0} />)).toJSON();
		expect(treeApp).toMatchSnapshot();
	});

	test('click on clear button', () => {
		jest.spyOn(navigator.clipboard, "readText").mockReturnValue(new Promise((resolve) => resolve("test")));
		render(renderWithRedux(<ContextMenu x={0} y={0} />));
		const btn = screen.getByTestId('clear');
		
		fireEvent.click(btn);
		expect(mockClearExpression).toBeCalledTimes(1);

	});

	test('click on past button', async () => {
		jest.spyOn(navigator.clipboard, "readText").mockReturnValue(new Promise((resolve) => resolve("test")));
		render(renderWithRedux(<ContextMenu x={0} y={0} />));
		const btn = screen.getByTestId('past');
		
		await fireEvent.click(btn);
		expect(mockPastExpression).toBeCalledTimes(1);

		await fireEvent.click(btn);
		expect(mockPastExpression).toBeCalledTimes(2);
	});

	test('click on past button with empty buffer', async () => {
		jest.spyOn(navigator.clipboard, "readText").mockReturnValue(new Promise((resolve) => resolve("")));
		render(renderWithRedux(<ContextMenu x={0} y={0} />));
		const btn = screen.getByTestId('past');
		
		await fireEvent.click(btn);
		expect(mockPastExpression).toBeCalledTimes(0);

		await fireEvent.click(btn);
		expect(mockPastExpression).toBeCalledTimes(0);
	});
});

