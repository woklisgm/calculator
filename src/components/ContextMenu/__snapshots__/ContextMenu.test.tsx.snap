// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard tests snapshot 1`] = `
<div
  className="contextmenu"
  style={
    Object {
      "left": 0,
      "top": 0,
    }
  }
>
  <div
    className="item"
    data-testid="past"
    onClick={[Function]}
  >
    Вставить
  </div>
  <div
    className="item"
    data-testid="clear"
    onClick={[Function]}
  >
    Очистить
  </div>
  <div
    className="item"
  >
    ...
  </div>
</div>
`;
